docs: OOM recovery in Redis best practices (#7194)
juicedata/juicefs: docs/en/administration/metadata/redis_best_practices.md

Among them, `used_memory_rss` is the total memory size actually used by Redis, which includes not only the size of data stored in Redis (that is, `used_memory_dataset` above) but also some Redis [system overhead](https://redis.io/commands/memory-stats) (that is, `used_memory_overhead` above). As mentioned earlier that the metadata of each file occupies about 300 bytes, this is actually calculated by `used_memory_dataset`. If you find that the metadata of a single file in your JuiceFS file system occupies much more than 300 bytes, you can try to run [`juicefs gc`](../../reference/command_reference.mdx#gc) command to clean up possible redundant data.

> ### Recover from out-of-memory (OOM) {#recover-from-oom}

Reviewer: ```suggestion
### Recover from out-of-memory (OOM) conditions {#recover-from-oom}
```

> Once Redis memory usage reaches the `maxmemory` limit, all write operations are rejected with the error `OOM command not allowed when used memory > 'maxmemory'`. At this point, even attempting to free up space by deleting files with [`juicefs rmr`](../../reference/command_reference.mdx#rmr) or by [purging the trash](../../security/trash.md#purge) will fail, because these deletion operations still need to write to Redis to complete.

Reviewer: ```suggestion
Once Redis memory usage reaches the `maxmemory` limit, all write operations are rejected with the error `OOM command not allowed when used memory > 'maxmemory'`. At this point, even attempts to free up space by deleting files with [`juicefs rmr`](../../reference/command_reference.mdx#rmr) or by [purging the trash](../../security/trash.md#purge) will fail, because these operations still require writes to Redis to complete.
```

> To recover from this state, consider the following approaches:

Reviewer: ```suggestion
To recover from this state, consider the following options:
```

> - **Temporarily raise the memory limit**: First use [`CONFIG SET maxmemory <new-value>`](https://redis.io/commands/config-set) to temporarily raise the Redis memory limit so that writes become available again, then run `juicefs rmr` or purge the trash to free up space, and finally restore `maxmemory` to its original value afterwards. If deletion alone cannot free up enough space, you can also use this window to scale up Redis.

Reviewer: ```suggestion
- **Temporarily increase the memory limit**: Use [`CONFIG SET maxmemory <new-value>`](https://redis.io/commands/config-set) to temporarily raise the Redis memory limit so writes become available again. Then run `juicefs rmr` or purge the trash to free up space and restore `maxmemory` to its original value afterward. If deletion alone cannot free enough space, you can also use this window to scale up Redis.
```

> - **Reserve a placeholder key (recommended preparation in advance)**: While the file system is running normally, write a large placeholder key into Redis ahead of time to reserve a portion of memory. When an OOM situation occurs, simply delete this placeholder key to immediately free up space, allowing `juicefs rmr` and trash-purging operations to proceed.

Reviewer: ```suggestion
- **Reserve a placeholder key (recommended preparation in advance)**: While the file system is running normally, write a large placeholder key into Redis ahead of time to reserve a portion of memory. If an OOM condition occurs, deleting this key immediately releases memory, allowing `juicefs rmr` and trash-purging operations to proceed.
```

> To avoid triggering OOM in the first place, take preventive measures in advance:

Reviewer: ```suggestion
To reduce the risk of OOM, take the following preventive measures:
```

> - Reserve some buffer for `maxmemory` when deploying, rather than setting it close to the physical memory limit, leaving room for operations such as deletions.

Reviewer: ```suggestion
- When deploying, reserve some buffer for `maxmemory`, rather than setting it close to the physical memory limit, leaving room for operations such as deletions.
```

> - When using a managed service, configure capacity monitoring and alerts so that you can scale up in time before memory usage approaches the limit.

Reviewer: ```suggestion
- If you are using a managed service, configure capacity monitoring and alerting so that you can scale up before memory usage approaches the limit.
```
